platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/linking/InOrderLinker.java

```diff
@@ -16,99 +16,161 @@
-import com.swirlds.logging.LogMarker;
+import com.swirlds.common.crypto.Hash;
+import com.swirlds.common.sequence.map.SequenceMap;
+import com.swirlds.common.sequence.map.StandardSequenceMap;
+import com.swirlds.common.system.events.BaseEventHashedData;
+import com.swirlds.common.system.events.EventDescriptor;
 import com.swirlds.platform.EventStrings;
 import com.swirlds.platform.event.GossipEvent;
+import com.swirlds.platform.gossip.IntakeEventCounter;
 import com.swirlds.platform.internal.EventImpl;
 import edu.umd.cs.findbugs.annotations.NonNull;
-import java.time.Duration;
+import java.util.HashMap;
+import java.util.Map;
 import java.util.Objects;
-import java.util.function.Function;
+import java.util.function.Consumer;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 /**
- * An {@link EventLinker} which expects events to be provided in topological order. If an
- * out-of-order event is provided, it is logged and discarded.
+ * Links events.
+ * <p>
+ * Expects events to be provided in topological order. If an out-of-order event is provided, it is logged and discarded.
+ * <p>
+ * Note: This class doesn't have a direct dependency on the {@link com.swirlds.platform.gossip.shadowgraph.ShadowGraph ShadowGraph},
+ * but it is dependent in the sense that the Shadowgraph is currently responsible for eventually unlinking events.
  */
-public class InOrderLinker extends AbstractEventLinker {
+public class InOrderLinker {
     private static final Logger logger = LogManager.getLogger(InOrderLinker.class);
-    private final ParentFinder parentFinder;
-    /** Provides the most recent event by the supplied creator ID */
-    private final Function<NodeId, EventImpl> mostRecentEvent;
 
-    private EventImpl linkedEvent = null;
+    /**
+     * The initial capacity of the {@link #parentDescriptorMap} and {@link #parentHashMap}
+     */
+    private static final int INITIAL_CAPACITY = 1024;
 
-    private final RateLimitedLogger unprocessedEventLogger;
-    private final RateLimitedLogger missingParentsLogger;
+    /**
+     * A sequence map from event descriptor to event.
+     * <p>
+     * The window of this map is shifted when the minimum generation non-ancient is changed, so that only non-ancient
+     * events are retained.
+     */
+    private final SequenceMap<EventDescriptor, EventImpl> parentDescriptorMap =
+            new StandardSequenceMap<>(0, INITIAL_CAPACITY, true, EventDescriptor::getGeneration);
 
+    /**
+     * A map from event hash to event.
+     * <p>
+     * This map is needed in addition to the sequence map, since we need to be able to look up parent events based on
+     * hash. Elements are removed from this map when the window of the sequence map is shifted.
+     */
+    private final Map<Hash, EventImpl> parentHashMap = new HashMap<>(INITIAL_CAPACITY);
+
+    /**
+     * Linked events are passed to this consumer.
+     */
+    private final Consumer<EventImpl> eventConsumer;
```

Comment:
Passing data onto the next consumer in line is a pattern we decided not to go with in the new wiring. The preference is that things work the way they used to, linked events are returned by the component.

Reply:
I've been temporarily continuing with this event consumer model in these components since it allows me to easily wire things up before having access to the new framework, to expose any logic flaws in the planned refactor.

This way even if the framework takes a while to finally merge, I can continue making progress